src/som/interpreter/actors/SPromise.java

```diff
@@ -328,8 +329,74 @@ public static void setSOMClass(final SClass cls) {
       resolverClass = cls;
     }
 
-    public final void onError() {
-      throw new NotYetImplementedException(); // TODO: implement
+    public static void onError(final SAbstractObject exception, final Object wrapped, final SPromise p, final Actor current) {
+      //System.out.println("called on error");
+      // for trace  TODO fix tracing
+      if (VmSettings.PROMISE_RESOLUTION) {
+        ActorExecutionTrace.promiseRuin(p.getPromiseId(), exception);
+      }
+      synchronized (wrapped) {
+        //System.out.println(p.chainedPromise);
+        //System.out.println(p.chainedPromiseExt);
+       handleOrPasstoChain(exception, wrapped, p, current);
+      }
+    }
+
+    public static void handleOrPasstoChain(final SAbstractObject exception, final Object wrapped, final SPromise p, final Actor current){
+      //System.out.println("called handler or pass to chain");
+      if (!runHandlersUnsync(exception, wrapped, p, current)){
+          //if I did not handle the exception ask chain to handle
+          ruinChainedPromisesUnsynch(exception, p, current);
+      }
+    }
+
+    private static boolean runHandlersUnsync(final SAbstractObject exception, final Object wrapped, final SPromise p, final Actor current) {
+        //sSystem.out.println("called run handlers");
+        synchronized (p) {
+          assert p.assertNotCompleted();
+          p.value = exception;
+          p.resolutionState = Resolution.ERRONEOUS;
+        }
+
+        boolean handled = false;
+        
+        //execute all exceptionHandlers if the type of wrapped is equal to the exceptionClass the handler handles.
+        if (p.onException != null) {
+          for (int i = 0; i < p.onException.size(); i++) {
+            SClass exceptionClass = p.onException.get(i);
```

Comment:
I was under the impression that we figured out that the `onException` stuff can't be supported because it implies visible race conditions. Or is this not the case after all?
I guess it depends on how this is realized. But from the sequential world, filtering would be the way to go, which is a race. So, it can't be done, I think.

I'd rather remove the `on:` exception support.